### ISSUE TITLE
[rush] Make API design more symmetrical between commandKind builk and commandKind global by adding shellCommand to bulk

### ISSUE THREAD

**Issue by gabrielliwerant** [microsoft/rushstack#2442]
## Summary

When switching a command-line.json `global` `commandKind` command to a `bulk` command, the purpose of the `name` field changes without explanation in the global section of the json, and it is unclear why we cannot separate `shellCommand` from `name` with `bulk` `commandKind` the way we can with `global`.

## Details

The json API for `commandKind` `bulk` couples together the `name` field with the actual shell command, unlike `commandKind` `global` which allows both a `shellCommand` and a `name` field. The purpose of this is not explained in the `global` section, but it seems logical to separate the name of the command from the actual command, as commands themselves might not have the best name, and `name` is something that normally indicates metadata.

In the case of the `bulk` `commandKind`, `name` is overloaded to provide both metadata and the shell command. This seems unnecessary and leads to confusion if switching `commandKind`s.

I think the API would be easier to reason about if `bulk` had its own `shellCommand` with the responsibility of holding the actual command, such that the metadata can remain separate between both `global` and `bulk` commands, simplifying the process of converting one to the other, and reducing confusion about what fields are needed and which are not.

## Standard questions

Please answer these questions to help us investigate your issue more quickly:

| Question | Answer |
| -------- | -------- |
| `@microsoft/rush` globally installed version? | 5.36.1 |
| `rushVersion` from rush.json? | 5.36.1 |
| `useWorkspaces` from rush.json? | false |
| Operating system? | Mac |
| Would you consider contributing a PR? | Yes |
| Node.js version (`node -v`)? | 14.15.4 |

**Comment by octogonz** [comment]
> When switching a command-line.json `"commandKind": "global"` command to a `"commandKind": "bulk"` command, the purpose of the `name` field changes without explanation in the global section of the json, and it is unclear why we cannot separate `shellCommand` from `name` with `"commandKind": "bulk"` the way we can with `"commandKind": "global"`.

## Current design 
This is not exactly true.  The official "purpose" of `name` is to determine the name of the Rush CLI command.   By "command" I mean `do-something` in this example command line:
```shell
$ rush do-something
```

For "global" commands, `rush do-something` invokes an arbitrary script specified by `shellCommand`.  The script gets invoked once.

For "bulk" commands, `rush do-something` looks in each project's **package.json** file to find an NPM script called `"do-something"`.  This script is executed separately for each project.

**Comment by octogonz** [comment]
> I think the API would be easier to reason about if `bulk` had its own `shellCommand` with the responsibility of holding the actual command, such that the metadata can remain separate between both `global` and `bulk` commands, simplifying the process of converting one to the other, and reducing confusion about what fields are needed and which are not.

With this proposed design, suppose that I want my bulk command to invoke the `"test"` script from package.json in each project folder.  How would I represent that?  

And how would the `ignoreMissingScript` setting  work?

**Comment by gabrielliwerant** [comment]
> With this proposed design, suppose that I want my bulk command to invoke the "test" script from package.json in each project folder. How would I represent that?

I agree, given the purpose of the `bulk` that it doesn't make sense to re-purpose `shellCommand`. Maybe it's just me, but the thing I think it would be good to replicate from `global` is the idea that the `name` field can act as an alias for a much more complicated command. With `bulk`, it seems to be that `name` is overloaded as both identifier for the command and the command to run, such that I can't name it one thing and specify a more complicated entry point into the script (e.g. `rush my-very-large-script-name:with-stuff-namespaced-in-script-file` to `rush foo`. I know that I can already simplify in this way within the script files, but that requires potentially altering the script files in many packages, whereas running a bulk command would be simpler.

Additionally, I think there's value from a schema perspective in keeping `name` and `some-command` separate, again for identification or aliasing purposes. So my new proposal would look something like one of the two scenarios:

```
    {
      "commandKind": "bulk",
      "name": "test",
      "scriptName": "test-the-things:thing-type"
    }
```

or

```
    {
      "commandKind": "bulk",
      "scriptName": "test-the-things:thing-type"
    }
```

> And how would the ignoreMissingScript setting work?

I believe `ignoreMissingScript` would work as before in the cases I suggest, as `bulk` would still be fundamentally about running a named script from the package files.

It's possible that this kind of a change is not worthwhile. I don't know how many others may have stumbled over the same confusion. For me, it took a while for my brain to grasp what was happening with the `name` field despite the documentation, possibly because I was continuing to think in terms of the `global` command. I understand the different between the two, and it makes sense, but I thought a change in the API here might save others the same trouble I had.

**Comment by gabrielliwerant** [comment]
> With this proposed design, suppose that I want my bulk command to invoke the "test" script from package.json in each project folder. How would I represent that?

I agree, given the purpose of the `bulk` that it doesn't make sense to re-purpose `shellCommand`. Maybe it's just me, but the thing I think it would be good to replicate from `global` is the idea that the `name` field can act as an alias for a much more complicated command. With `bulk`, it seems to be that `name` is overloaded as both identifier for the command and the command to run, such that I can't name it one thing and specify a more complicated entry point into the script (e.g. `rush my-very-large-script-name:with-stuff-namespaced-in-script-file` to `rush foo`. I know that I can already simplify in this way within the script files, but that requires potentially altering the script files in many packages, whereas running a bulk command would be simpler.

Additionally, I think there's value from a schema perspective in keeping `name` and `some-command` separate, again for identification or aliasing purposes. So my new proposal would look something like one of the two scenarios:

```
    {
      "commandKind": "bulk",
      "name": "test",
      "scriptName": "test-the-things:thing-type"
    }
```

or

```
    {
      "commandKind": "bulk",
      "scriptName": "test-the-things:thing-type"
    }
```

> And how would the ignoreMissingScript setting work?

I believe `ignoreMissingScript` would work as before in the cases I suggest, as `bulk` would still be fundamentally about running a named script from the package files.

For me, it took a while for my brain to grasp what was happening with the `name` field despite the documentation, possibly because I was continuing to think in terms of the `global` command. I understand the different between the two now, and it makes sense, but I thought a change in the API here might save others the same trouble I had.

**Comment by benkeen** [comment]
Good thread. I'd also like more control over how I name the bulk commands to keep them separate from the actual commands ran in the individual packages. In context within a single package, some long-winded script name often makes sense. But running globally is a different context & a briefer/different command name would be better.
